pkg/repository/prisma/buffered.go

```diff
@@ -0,0 +1,276 @@
+package prisma
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/errgroup"
+)
+
+type ingestBufState int
+
+const (
+	started ingestBufState = iota
+	initialized
+)
+
+// e.g. T is eventOpts and U is *dbsqlc.Event
+
+type IngestBuf[T any, U any] struct {
+	outputFunc func(ctx context.Context, items []T) ([]U, error)
+	sizeFunc   func(T) int
+
+	state       ingestBufState
+	maxCapacity int           // max number of items to hold in buffer before we flush
+	flushPeriod time.Duration // max time to hold items in buffer before we flush
+
+	inputChan          chan *inputWrapper[T, U]
+	lastFlush          time.Time
+	internalArr        []*inputWrapper[T, U]
+	sizeOfData         int // size of data in buffer
+	maxDataSizeInQueue int // max number of bytes to hold in buffer before we flush
+
+	l      *zerolog.Logger
+	lock   sync.Mutex
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type inputWrapper[T any, U any] struct {
+	item     T
+	doneChan chan<- *flushResponse[U]
+}
+
+type IngestBufOpts[T any, U any] struct {
+	maxCapacity        int
+	flushPeriod        time.Duration
+	maxDataSizeInQueue int
+	outputFunc         func(ctx context.Context, items []T) ([]U, error)
+	sizeFunc           func(T) int
+	l                  *zerolog.Logger
+}
+
+// NewIngestBuffer creates a new buffer for any type T
+func NewIngestBuffer[T any, U any](opts IngestBufOpts[T, U]) *IngestBuf[T, U] {
+
+	inputChannelSize := opts.maxCapacity
+	if inputChannelSize < 100 {
+		inputChannelSize = 100
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &IngestBuf[T, U]{
+		state:              initialized,
+		maxCapacity:        opts.maxCapacity,
+		flushPeriod:        opts.flushPeriod,
+		inputChan:          make(chan *inputWrapper[T, U], inputChannelSize),
+		lastFlush:          time.Now(),
+		internalArr:        make([]*inputWrapper[T, U], 0),
+		sizeOfData:         0,
+		maxDataSizeInQueue: opts.maxDataSizeInQueue,
+		outputFunc:         opts.outputFunc,
+		sizeFunc:           opts.sizeFunc,
+		l:                  opts.l,
+		ctx:                ctx,
+		cancel:             cancel,
+	}
+}
+
+func (b *IngestBuf[T, U]) validate() error {
```

Comment:
nit we generally validate with struct tags:

```
type IngestBufOpts[T any, U any] struct {
    MaxCapacity        int                                                `validate:"required,gt=0"`
    FlushPeriod        time.Duration                                      `validate:"required,gt=0"`
    MaxDataSizeInQueue int                                                `validate:"required,gt=0"`
    OutputFunc         func(ctx context.Context, items []T) ([]U, error)  `validate:"required"`
    SizeFunc           func(T) int                                        `validate:"required"`
    Logger             *zerolog.Logger                                    `validate:"required"`
}
```

Reply:
done